### PR TITLE
Parse pbc from/to structure-node

### PR DESCRIPTION
**Pull request by hdsassnick**
Dear @yakutovicha ,

I would like to suggest another change in the code.
I noticed that the boundary conditions are not considered when the output structure is parsed, 
which can lead to wrong boundary conditions in the output structure.
For example, the seekpathtool in aiida-core needs 3D periodic boundary conditions to properly work.
But also, in general, I think it is a good idea to make use of this feature of the aiida-structure node.

So, I adapted the Cp2kCalculation to put the boundary conditions into the &CELL section of the input file and 
the function parse_cp2k_trajectory is changed such that it retrieves the periodic boundary conditions and adds 
it to the ase-Atoms object.

I had to adapt two tests in order to make it work:
* example_geopt.py: Here, the pbc was inconsistent with the Poisson-Solver (default is XYZ in cp2k). Now, the calculation
	runs with non-periodic boundary conditions. 
	As such, the total energy and bond length slightly changed and needed to be adapted.
* example_precsion.py: The ase-Atoms object was having non-periodic boundary conditions, overriding the default pbc in cp2k.
	Here, I set the boundary conditions to True in the ase-Atoms object to fix the test.

Please let me know what you think.

Have a nice day.
Best wishes,
Holger

### CODE REVIEW

**Review by yakutovicha — COMMENTED** [review]
Hey, @hdsassnick thanks.

While I agree that the output parser should parse the periodicity of the structure, I am not sure the same should be done for the input generator.

There is a number of reasons for that:

1. `FORCE_EVAL/SUBSYS/CELL/PERIODIC` is not the only place where you can define periodicity - you can also specify it in the Poisson section (`FORCE_EVAL/DFT/POISSON/PERIODIC`) and in some other places.
2. We also need to make sure to make that things are backwards compatible: the previous results might become different as before the only source of truth was the input dictionary, while in this PR it is the structure and the input dictionary.

**Comment by hdsassnick** [comment]
Hi @yakutovicha ,

Thanks so much for having a look at the pull request and for your detailed comments.

I am not sure I understand your concerns about the input generator.

1. Concerning the "Poisson" section (e.g. for DFT or MM), I left this section out because the choice goes along with the Poisson-solver that is applied. This choice is in some cases not trivial and in my opinion should be done by the user.
If we set the pbc there too, it also might clash with the chosen Poisson-Solver. However, I don't see any harm in setting the pbc in the CELL section since it can be overwritten by the input-dictionary anyway. Also, in my opinion it is intuitive that this information is used from the structure node and not left as "default".
Another feature could be to check if `FORCE_EVAL/SUBSYS/CELL/PBC` is consistent with `FORCE_EVAL/DFT/POISSON/PBC` 
or `FORCE_EVAL/MM/POISSON/PBC` and give a warning in case the keywords are inconsistent.
For QMMM simulations, however, this would be a bit tricky ...

2. I agree that it destroys backwards compatibility as demonstrated with the examples.
But as I see it the source of the results was already before the structure (atomic positions/cell vectors) + 
input dictionary (calculation parameters) in the first place?
The only difference is that in case the pbc is not set in the dictionary, it is retrieved from the structure.

What do you think?

Best wishes,
Holger

**Comment by yakutovicha** [comment]
> 1. Another feature could be to check if `FORCE_EVAL/SUBSYS/CELL/PBC` is consistent with `FORCE_EVAL/DFT/POISSON/PBC`
>    or `FORCE_EVAL/MM/POISSON/PBC` and give a warning in case the keywords are inconsistent.

That is why, essentially, I am so hesitant to move forward with this PR (the input part of it). Because doing this feels like a huge mess. And, ultimately, we should let the user break things the way they want, the plugin should not check the science.

However, I agree that this feature is handy as soon as it doesn't introduce more problems.

@cpignedoli, @eimrek - since you also actively using the plugin, could tell us your opinion on the topic? Should we take the periodicity from the structure file (keeping the input dictionary with a higher priority) or should we extract this information _only_ from the parameter dictionary?

**Comment by eimrek** [comment]
hi!

The output parsing makes perfect sense.

Regarding the input generation, from the point of view of consistency, this also makes sense.

However, what might be slightly problematic is that the user might be relying on the default cp2k behavior (leaving the periodicity explicitly unspecified) and then the structure might introduce `PERIODIC NONE`, if you're not paying attention to structure's tags. Especially e.g. with `ase` the pbc by default is set to false in all directions. This might be a rare case, however. But I think there is some benefit in being able to be certain that what you set or leave unset in the `parameters` dictionary also ends in the final input script. Also considering that when using `PERIODIC NONE`, in most cases you would have to manually handle other accompanying sections anyways (e.g. the poisson solver.)

Overall, I am slightly leaning towards keeping the old convention (if `periodic` is unset in the `parameters`, then it's also unset in the final script), but I'd also be fine with the change. @hdsassnick do you have a specific use-case in mind, which is clearly improved by this change?

**Comment by hdsassnick** [comment]
Hi @eimrek ,

Thank you for the helpful comment.

I have to admit that the practical application was about parsing the pbc into the output-structure.
While implementing the feature I thought that it would make more sense to parse it from the structure into the input-file as well.
The motivation was that the section `&CELL` actually only consists of keywords that are defined by the input-structure or are related to it.

But I also see your point. At the end practicability should have priority. 

Have a nice weekend.
Best wishes,
Holger

**Comment by hdsassnick** [comment]
Hi,

since there were some concerns about the input-generator I removed the pbc-parsing there. I think this is a good compromise where we can all be on the same page that the new feature adds to the functionality of the package without breaking downward compatibility for the input.

@yakutovicha could you please have another look at the pull request?
Should we also change/add some tests for this feature?

All the best,
Holger

File: aiida_cp2k/utils/parser.py [hunk 1]
```diff
@@ -391,4 +391,11 @@ def parse_cp2k_trajectory(content):
     cell_str = [line[1:] for line in cell_lines if line[0] in "ABC"]
     cell = np.array(cell_str, np.float64)
 
-    return {"symbols": symbols, "positions": positions, "cell": cell, "tags": tags}
+    # parse periodic boundary conditions
+    cell_pbc = [True, True, True]  # In case keyword is not set: Default in cp2k is XYZ
+    for line in cell_lines:
+        if line[0] == 'PERIODIC':
+            cell_pbc_str = line[-1]
+            cell_pbc = [(dir in cell_pbc_str) for dir in ['X', 'Y', 'Z']]
```
Comment (yakutovicha) [inline aiida_cp2k/utils/parser.py:399]:
```suggestion
    # parse periodic boundary conditions
    pbc_str = [ for line in cell_lines if 'PERIODIC' in line ]
    cell_pbc = [(dir in pbc_str) for dir in ['X', 'Y', 'Z']] if len(pbc_str) == 1 else [True, True, True]
```

**Review by yakutovicha — COMMENTED** [review]
Thanks, @hdsassnick.

I proposed a small change to the PBC parser. The idea is essentially the same, but I was trying to make it a bit shorter and clearer. However, I am not sure if I achieved that, so I leave it up to you to chose which implementation you prefer.

Concerning the tests: absolutely. Please add them to https://github.com/aiidateam/aiida-cp2k/tree/develop/test. The restart file example should go into: https://github.com/aiidateam/aiida-cp2k/tree/develop/test/outputs. You might want to add two tests: when `PERIODIC` is present and when it is not present.

**Comment by hdsassnick** [comment]
Hi @yakutovicha ,

sorry for taking so long. I added a test for the parsing.
I left the structure of the code for now. I am fine either way, as long as the code works.

What do you think?

Best wishes,
Holger

**Review by yakutovicha — APPROVED** [review]
LGTM! Thanks a lot @hdsassnick!
